### PR TITLE
Training: Add Fine-Tune API Docs

### PR DESCRIPTION
**Pull request by andreyvelich**
Related: https://github.com/kubeflow/training-operator/issues/2013
This is draft PR for our new Fine-Tune API in Kubeflow Training Operator.
We will work on the page structure in this Google doc to finalise it: https://docs.google.com/document/d/18PuuaDRISj5mlrBn1GJrxwuB6Z5zTtXKpVbLUIeLx-8/edit?usp=sharing.

### CODE REVIEW

**Comment by andreyvelich** [comment]
I added content from the Google doc and one tutorial.
Please let me know what do you think.
/assign @StefanoFioravanzo @kubeflow/wg-training-leads @deepanker13 @kuizhiqing

**Review by hbelmiro — COMMENTED** [review]
/lgtm

File: content/en/docs/components/training/user-guides/fine-tuning.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
++++
+title = "How to Fine-Tune LLM with Kubeflow"
```
Comment (StefanoFioravanzo) [inline content/en/docs/components/training/user-guides/fine-tuning.md:2]:
```suggestion
title = "How to Fine-Tune LLMs with Kubeflow"
```

File: content/en/docs/components/training/user-guides/fine-tuning.md [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+fine-tuning tasks often require extensive manual intervention, including the configuration of
+training environments and the distribution of data across nodes. The Fine Tune APIs aim to simplify
+this process, offering an easy-to-use Python interface that abstracts away the complexity involved
+in setting up and executing fine-tuning tasks on distributed systems.
+
+## How to use Fine-Tuning API ?
+
+[Training Operator Python SDK](/docs/components/training/installation/#installing-training-python-sdk)
+implements a [`train` Python API](https://github.com/kubeflow/training-operator/blob/6ce4d57d699a76c3d043917bd0902c931f14080f/sdk/python/kubeflow/training/api/training_client.py#L112)
+that simplify ability to fine-tune LLMs with distributed PyTorchJob workers.
```
Comment (StefanoFioravanzo) [inline content/en/docs/components/training/user-guides/fine-tuning.md:36]:
```suggestion
that simplifies the ability to fine-tune LLMs with distributed PyTorchJob workers.
```

File: content/en/docs/components/training/user-guides/fine-tuning.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+You need to provide the following parameters to use `train` API:
```
Comment (StefanoFioravanzo) [inline content/en/docs/components/training/user-guides/fine-tuning.md:38]:
```suggestion
You need to provide the following parameters to use the `train` API:
```

File: content/en/docs/components/training/user-guides/fine-tuning.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    num_procs_per_worker=2, # nproc-per-node parameter for torchrun command.
+    resources_per_worker={
+        "gpu": 2,
+        "cpu": 5,
+        "memory": "10G",
+    },
+)
+```
+
+After you execute `train` API, Training Operator will orchestrate appropriate PyTorchJob resources
```
Comment (StefanoFioravanzo) [inline content/en/docs/components/training/user-guides/fine-tuning.md:99]:
```suggestion
After you execute `train`, Training Operator will orchestrate appropriate PyTorchJob resources
```

File: content/en/docs/components/training/user-guides/fine-tuning.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+- Pre-trained model parameters.
+- Dataset parameters.
+- Trainer parameters.
+- Number of PyTorch workers and resources per workers.
+
+For example, you can use `train` API as follows to fine-tune BERT model using Yelp Review dataset
+from HuggingFace Hub:
+
+```python
```
Comment (StefanoFioravanzo) [inline content/en/docs/components/training/user-guides/fine-tuning.md:48]: If I copy paste this snippet into a notebook, does it run seamlessly? What are the required dependencies? Do we need to provide a `pip install` command to make sure that this snippet runs? Also, what is the expected output?

File: content/en/docs/components/training/user-guides/fine-tuning.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+        "cpu": 5,
+        "memory": "10G",
+    },
+)
+```
+
+After you execute `train` API, Training Operator will orchestrate appropriate PyTorchJob resources
+to fine-tune LLM.
+
+## Architecture
```
Comment (StefanoFioravanzo) [inline content/en/docs/components/training/user-guides/fine-tuning.md:102]: This should go to a "Reference"

File: content/en/docs/components/training/user-guides/fine-tuning.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+### LLM Trainer
+
+Trainer implements training loop to fine-tune LLM. Currently, Training Operator supports
+[HuggingFace trainer](https://github.com/kubeflow/training-operator/blob/6ce4d57d699a76c3d043917bd0902c931f14080f/sdk/python/kubeflow/trainer/hf_llm_training.py#L118-L139)
+to fine-tune LLMs.
+
+You can implement your own trainer for other ML use-cases such as image classification,
+voice recognition, etc.
+
+## User Value for this Feature
```
Comment (StefanoFioravanzo) [inline content/en/docs/components/training/user-guides/fine-tuning.md:151]: I think we can just fold this under `Why Training Operator Fine-Tune API Matter ?` by stripping the title `User Value for this Feature`

File: content/en/docs/components/training/user-guides/fine-tuning.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+In the rapidly evolving landscape of machine learning (ML) and artificial intelligence (AI),
+the ability to fine-tune pre-trained models represents a significant leap towards achieving custom
+solutions with less effort and time. Fine-tuning allows practitioners to adapt large language models
+(LLMs) like BERT or GPT to their specific needs by training these models on custom datasets.
+This process maintains the model's architecture and learned parameters while making it more relevant
+to particular applications. Whether you're working in natural language processing (NLP),
+image classification, or another ML domain, fine-tuning can drastically improve performance and
+applicability of pre-existing models to new datasets and problems.
+
+## Why Training Operator Fine-Tune API Matter ?
```
Comment (StefanoFioravanzo) [inline content/en/docs/components/training/user-guides/fine-tuning.md:22]: I feel like this is out of place here. The how-to guide provides a step by step sequenced guide on how to achieve a very specific task. A how-to guide generally does not provide Reference or Explanation. It seems to me we are writing some paragraphs that would be more suited to an "Explanation" section. This is the fourth content types proposed by Diataxis - see here https://diataxis.fr/explanation/

I can very well see a page under "Explanation" titled "LLM Fine-Tune APIs in Kubeflow" where we discuss why we need it and how it fits into the ecosystem. Basically what you wrote already, plus a little bit of refactoring. WDYT?

File: content/en/docs/components/training/user-guides/fine-tuning.md (hunk 5 above)
Comment (andreyvelich) [reply content/en/docs/components/training/user-guides/fine-tuning.md:48]: Let me add the prerequisites to run this API.

**Comment by andreyvelich** [comment]
I addressed your comments @StefanoFioravanzo.
Regarding this comment:

> Also, what is the expected output?

How we can show the expected output ? Our LLM trainer doesn't support any output yet: https://github.com/kubeflow/training-operator/blob/master/sdk/python/kubeflow/trainer/hf_llm_training.py#L178, so we need to work in the future to understand how user should consume the fine-tuned model.
E.g. exporting to S3 or to other storage.
cc @johnugeorge @deepanker13

File: content/en/docs/components/training/user-guides/fine-tuning.md (hunk 8 above)
Comment (StefanoFioravanzo) [reply content/en/docs/components/training/user-guides/fine-tuning.md:22]: That's a very good question. In the how to guide, we can have something like: "If you want to learn more about how the fine tune API fit in the Kubeflow ecosystem, head to <...>".

And in the exlanation guide, we can say something like: "Head to <link> for a quick start tutorial on using LLM Fine-tune APIs. Head to <link> for a reference architecture on the control plane implementation"

And generally we can have links to how-tos in tutorials and reference guides. So in general, let's try to link related topics together when it makes sense for a user to follow that train of thought

**Comment by StefanoFioravanzo** [comment]
> so we need to work in the future to understand how user should consume the fine-tuned model.

Issue + KF 1.10 tag? :)

File: content/en/docs/components/training/user-guides/fine-tuning.md (hunk 8 above)
Comment (andreyvelich) [reply content/en/docs/components/training/user-guides/fine-tuning.md]: Sure, what do you think about it @StefanoFioravanzo ?
https://github.com/kubeflow/website/pull/3718/commits/7d30f127d35cc9a306328ccc4d309b7b66c9c54f

Comment (StefanoFioravanzo) [reply content/en/docs/components/training/user-guides/fine-tuning.md]: Looks great!

**Comment by StefanoFioravanzo** [comment]
@andreyvelich  yes it does thank you!

**Review by tenzen-y — COMMENTED** [review]
Awsome documentation! Thank you!
/lgtm

**Comment by deepanker13** [comment]
@andreyvelich the links in fine-tuning.md are giving 404 page not found. Am I missing something?

**Comment by andreyvelich** [comment]
> @andreyvelich the links in fine-tuning.md are giving 404 page not found. Am I missing something?

@deepanker13 Did you check these links via Website preview: https://deploy-preview-3718--competent-brattain-de2d6d.netlify.app/ ?

**Comment by deepanker13** [comment]
> > @andreyvelich the links in fine-tuning.md are giving 404 page not found. Am I missing something?
> 
> @deepanker13 Did you check these links via Website preview: https://deploy-preview-3718--competent-brattain-de2d6d.netlify.app/ ?

@andreyvelich it's working with the preview. Thanks for the awesome documentation!
/lgtm

**Comment by StefanoFioravanzo** [comment]
@andreyvelich shall we merge this one?

**Pull request merged**